README.rst
colormaps
=========

Introduction
------------

Transform data into rgba data using discrete lookup tables or color
gradients.

Usage
-----

Create a discrete colormap with the same two colors::
    
    >>> import colormaps
    >>> colormap = colormaps.DiscreteColormap(
    ...     data=[(0, (255, 0, 0, 255)),         
    ...           (1, (0, 0, 255, 127))],
    ...     masked=(255, 0, 0, 255),
    ...     invalid=(0, 255, 0, 255),
    ... )

Create a gradient colormap from red to semi-transparent blue::
    
    >>> import colormaps
    >>> colormap = colormaps.GradientColormap(
    ...     data=[(0, (255, 0,   0, 255)),
    ...           (1, (  0, 0, 255, 127))],
    ...     size=2048,
    ...     masked=(0, 0, 255, 255),
    ... )                                     

Where size is the amount of entries in the prepared look-up table.

Calculate the value at some positions::

    >>> colormap([0, 0.5, 1])
    array([[255,   0,   0, 255],
           [127,   0, 127, 190],
           [  0,   0, 255, 127]], dtype=uint8)


Register the colormap for use in other places::

    >>> colormap.register('my_gradient')
    >>> colormaps.get('my_gradient')
    <GradientColormap: size 2048, limits 0 - 1, log False, interp False>

Use a collection of predefined colors using a manager::
    
    >>> manager = colormaps.Manager()
    >>> manager.get('jet')
    <GradientColormap: size 256, limits 0.0 - 1.0, log False, interp False>

    >>> sorted(manager.registered.keys())[:3]
    ['Accent', 'Accent_r', 'Blues']

Or use your own collection::

    >>> manager = colormaps.Manager('path/to/my/colormap/collection')

The path should point to a folder with json-files ('my-colormap.json')
containing the kwargs for colormaps.create() like this::

    {
      "type": "GradientColormap",
      "data": [
          [0.000, [  0,   0,   0, 255]],
          [0.700, [255, 255, 255, 0]],
          [1.000, [255, 255, 255, 255]]
      ]
    }


Development installation
------------------------

For development, you can use a docker-compose setup::

    $ docker-compose build --build-arg uid=`id -u` --build-arg gid=`id -g` lib
    $ docker-compose up --no-start
    $ docker-compose start
    $ docker-compose exec lib bash

Create a virtualenv::

    (docker)$ virtualenv .venv  # note that Dockerfile prepends .venv/bin to $PATH

Install stuff and run the tests::

    (docker)$ pip install -r requirements.txt
    (docker)$ pip install -e .[test]
    (docker)$ pytest

Update packages::
    
    (docker)$ rm -rf .venv
    (docker)$ virtualenv .venv
    (docker)$ pip install -e .
    (docker)$ pip freeze | grep -v colormaps > requirements.txt


CHANGES.rst
Changelog of colormaps
===================================================


1.8.8 (unreleased)
------------------

- Added optional validation to colormap constructors.


1.8.7 (2022-09-09)
------------------

- Fixed numpy deprecation warning.


1.8.6 (2021-01-20)
------------------

- Fixed test.


1.8.5 (2021-01-20)
------------------

- Mask values less or equal to zero for logarithmic colormaps (i.e. make them
  transparent). A user should still provide no or valid limits. Specifying
  limits outside the log domain will result in a runtime error.


1.8.4 (2020-07-14)
------------------

- Ensure slice indices are integers (DiscreteColormap).


1.8.3 (2020-01-28)
------------------

- Add test for floatcapabilities in discrete colormaps.

- Use int64 for indexing. Its faster.


1.8.2 (2019-09-03)
------------------

- Revert to numpy 1.16.4.

- Handle NaN values in colormaps.


1.8.1 (2019-07-16)
------------------

- Bump numpy to 1.17.0rc1. Indeed.


1.8 (2019-03-07)
----------------

- Solve an IndexError in some edge case in GradientColorMap.


1.7 (2019-01-16)
----------------

- Removed np.round() in gradient CM index calculation.


1.6 (2019-01-14)
----------------

- Performing additional masking in Data.convert()


1.5 (2018-03-08)
----------------

- Colormaps accept raster-store-like data dictionaries for better performance

- Mappings using the `interp` argument can handle unsorted values

- `interp` can handle mappings that change the sign (negate)

- The order of interp and log is swapped: first interp, then log


1.4.1 (2017-11-02)
------------------

- Fixed legend bug.


1.4 (2017-10-03)
----------------

- Cast integer interpolations to floats to fix test

- Be clearer about the in-place log() on the interpolation inputs

- Add function to give a range of numbers to use in legends


1.3 (2016-06-14)
----------------

- Add colormaps and labels for forests, deforestation and floods.

- Remove the white haze in radar images.

- Add Port Elisabeth SA soil and land-cover colormaps.

- Change drought to non-free colormap.

- Update labels and colors of indonesia flood map.

- Coerce to dicts for embedded labels in colormap definitions.

- Update MANIFEST file.


1.2.4 (2015-01-27)
------------------

- Add vietnam cover colors.


1.2.3 (2015-01-19)
------------------

- Add the missing 5min variant for radar.


1.2.2 (2015-01-15)
------------------

- Adjust base colormap for radar, differentiate as well.


1.2.1 (2014-12-23)
------------------

- Labeling is vectorized now.

- Add world soil colors and labels.


1.2 (2014-12-17)
----------------

- Add label functionality by locale


1.1.1 (2014-12-15)
------------------

- Rearrange some colors.


1.1 (2014-12-08)
----------------

- Really use masked for discrete when outside limits

- Add manager

- Add colormaps


1.0 (2014-11-25)
----------------

- Handle masked arrays

- Init from a config

- Do (log) normalization and interpolation

- Handle limits for both types of maps

- Custom values for invalid (only discrete) and masked

- Add cdict converter


0.1.2 (2014-05-22)
------------------

- Simplify access to registered colormaps


0.1.1 (2014-05-22)
------------------

- Any gradient stops are now allowed.


0.1 (2014-04-16)
----------------

- Initial project structure created with nensskel 1.34.dev0.

- Implemented gradients and discrete colormaps.


colormaps/tests.py
# -*- coding: utf-8 -*-
# (c) Nelen & Schuurmans, see LICENSE.rst.

from math import exp
import os
import shutil
import tempfile
import unittest

from pytest import raises
import numpy as np

from colormaps import DiscreteColormap
from colormaps import GradientColormap
from colormaps import core
from colormaps import utils
import colormaps

MASKED = [0, 0, 255, 255]
INVALID = [0, 255, 0, 255]
BLACK = [0, 0, 0, 255]


def discrete():
    colormap = {
        'type': 'DiscreteColormap',
        'masked': MASKED,
        'invalid': INVALID,
        'data': [
            (0, (127, 000, 000, 255)),
            (2, (255, 000, 000, 255)),
        ],
        'labels': {'en_EN': {0: 'label0', 2: 'label2'}}
    }
    return colormaps.create(colormap)


def gradient(size=3, log=False, free=True, interp=None,
             data_values=(3, 5)):
    colormap = {
        'type': 'GradientColormap',
        'size': size,
        'free': free,
        'log': log,
        'interp': interp,
        'masked': MASKED,
        'data': [
            (data_values[0], (127, 000, 000, 255)),
            (data_values[1], (255, 000, 000, 255)),
        ],
    }
    return colormaps.create(colormap)


def cdict():
    return {
        'blue': (
            (0.000, 0.5, 0.5),
            (0.110, 1.0, 1.0),
            (0.340, 1.0, 1.0),
            (0.650, 0.0, 0.0),
            (1.000, 0.0, 0.0)
        ),
        'green': (
            (0.000, 0.0, 0.0),
            (0.125, 0.0, 0.0),
            (0.375, 1.0, 1.0),
            (0.640, 1.0, 1.0),
            (0.910, 0.0, 0.0),
            (1.000, 0.0, 0.0)
        ),
        'red': (
            (0.000, 0.0, 0.0),
            (0.350, 0.0, 0.0),
            (0.660, 1.0, 1.0),
            (0.890, 1.0, 1.0),
            (1.000, 0.5, 0.5)
        )
    }


class TestColormap(unittest.TestCase):
    def test_gradient_repr(self):
        colormap = gradient()
        self.assertTrue(repr(colormap))

    def test_gradient_size(self):
        colormap = gradient(size=5)
        self.assertEqual(len(colormap), 5)

    def test_gradient_clip(self):
        colormap = gradient()
        self.assertEqual(
            colormap(7, limits=(7, 9)).tolist(),
            [127, 000, 000, 255],
        )

    def test_gradient_nan(self):
        colormap = gradient()
        self.assertEqual(colormap([np.nan]).tolist(), [[0, 0, 255, 255]])

    def test_gradient_interp(self):
        # piecewise interp
        colormap = gradient(interp=[(3, 0), (3.1, 0.5), (5, 1)])
        self.assertEqual(
            colormap([2, 3, 3.1, 5, 7]).tolist(),
            [[127, 000, 000, 255],
             [127, 000, 000, 255],
             [191, 000, 000, 255],
             [255, 000, 000, 255],
             [255, 000, 000, 255]],
        )

        # negate interp
        colormap = gradient(data_values=[-3, -5],
                            interp=[(1e8, -1e8), (-1e8, 1e8)])
        self.assertEqual(
            colormap([-2, -3, -4, -5, -6]).tolist(),
            [[127, 000, 000, 255],
             [127, 000, 000, 255],
             [191, 000, 000, 255],
             [255, 000, 000, 255],
             [255, 000, 000, 255]],
        )

    def test_gradient_interp_and_clip(self):
        colormap = gradient(interp=[(3, 0), (5, 1)])
        self.assertEqual(
            colormap(7, limits=(4, 6)).tolist(),
            [255, 000, 000, 255],
        )

    def test_gradient_non_free(self):
        colormap = gradient(free=False)
        self.assertEqual(
            colormap(7).tolist(),
            [255, 000, 000, 255],
        )

    def test_gradient_log(self):
        colormap = gradient(log=True, data_values=np.exp([3, 5]))
        self.assertEqual(
            colormap(np.exp([3, 4, 5])).tolist(),
            [[127, 000, 000, 255],
             [191, 000, 000, 255],
             [255, 000, 000, 255]],
        )

    def test_gradient_log_limits(self):
        colormap = gradient(log=True, data_values=np.exp([3, 5]))
        self.assertEqual(
            colormap(np.exp([3, 4, 5]), limits=np.exp([4, 5])).tolist(),
            [[127, 000, 000, 255],
             [127, 000, 000, 255],
             [255, 000, 000, 255]],
        )

    def test_gradient_log_interp(self):
        # shift interp
        colormap = gradient(log=True, interp=([-100, 0], [10000, 10100]),
                            data_values=np.exp([3, 5]) - 100)
        self.assertEqual(
            colormap(np.exp([3, 4, 5]) - 100).tolist(),
            [[000, 000, 255, 255],
             [000, 000, 255, 255],
             [191, 000, 000, 255]],
        )

        # negate interp
        colormap = gradient(log=True, interp=([-1e8, 1e8], [1e8, -1e8]),
                            data_values=-np.exp([3, 5]))
        self.assertEqual(
            colormap(-np.exp([3, 4, 5])).tolist(),
            [[000, 000, 255, 255],
             [000, 000, 255, 255],
             [000, 000, 255, 255]],
        )

    def test_gradient_legend_data_limits_doesnt_clip(self):
        colormap = gradient()
        # Should map the data to the 0...2 range, even if the limits of
        # the gradient are 3 and 5 (because that's irrelevant)
        self.assertEqual(
            colormap.get_legend_data([0, 2], 5).tolist(),
            [0, 0.5, 1, 1.5, 2]
        )

    def test_gradient(self):
        colormap = gradient()
        # scalars
        m = np.ma.masked
        self.assertEqual(colormap(m).tolist(), MASKED)
        self.assertEqual(colormap(7).tolist(), [191, 000, 000, 255])
        # dict (Store.get_data return value, fastest)
        self.assertEqual(
            colormap(dict(values=[2, 3, 4, 6], no_data_value=3)).tolist(),
            [[127, 000, 000, 255],
             MASKED,
             [191, 000, 000, 255],
             [255, 000, 000, 255]]
        )
        # MaskedArray
        self.assertEqual(
            colormap(np.ma.masked_equal([2, 3, 4, 6], 3)).tolist(),
            [[127, 000, 000, 255],
             MASKED,
             [191, 000, 000, 255],
             [255, 000, 000, 255]]
        )

    def test_gradient_no_labels(self):
        colormap = gradient()
        self.assertEqual(colormap.label([5, 6]), [5, 6])

    def test_gradient_legend_data(self):
        colormap = gradient()
        self.assertEqual(
            colormap.get_legend_data(limits=None, steps=3).tolist(),
            [3, 4, 5],
        )

    def test_gradient_legend_data_limits(self):
        colormap = gradient()
        self.assertEqual(
            colormap.get_legend_data(limits=(3, 5), steps=3).tolist(),
            [3, 4, 5],
        )

    def test_gradient_log_legend_data(self):
        colormap = gradient(log=True)
        limits = exp(1.2), exp(1.4)
        self.assertAlmostEqual(
            colormap.get_legend_data(limits=limits, steps=3).tolist()[1],
            exp(1.3),
        )

    def test_gradient_interp_legend_data(self):
        colormap = gradient(interp=[(3, 0), (5, 1)])
        self.assertEqual(
            colormap.get_legend_data(limits=(3, 5), steps=3).tolist(),
            [3, 4, 5],
        )

    def test_discrete_legend_data(self):
        colormap = discrete()
        # There are two indices, 0 and 2; without limits, they should all be
        # returned.
        self.assertEqual(
            colormap.get_legend_data(None, None).tolist(),
            [0, 2]
        )

    def test_discrete_legend_data_with_limits(self):
        colormap = discrete()
        # If we give limits, only indices in those (inclusive) are returned
        self.assertEqual(
            colormap.get_legend_data([1, 2], None).tolist(),
            [2]
        )

    def test_discrete_repr(self):
        colormap = discrete()
        self.assertTrue(repr(colormap))

    def test_discrete_clip(self):
        colormap = discrete()
        self.assertEqual(
            colormap(0, limits=(1, 2)).tolist(),
            MASKED,
        )

    def test_discrete_label(self):
        colormap = discrete()
        self.assertEqual(colormap.label(0), 'label0')
        self.assertEqual(colormap.label([0, 1, 2]), ['label0', 1, 'label2'])

    def test_discrete(self):
        colormap = discrete()
        # scalar
        self.assertEqual(colormap(np.ma.masked).tolist(), MASKED)
        self.assertEqual(colormap(5).tolist(), MASKED)
        self.assertEqual(colormap(2).tolist(), [255, 000, 000, 255])
        # dict (Store.get_data return value, fastest)
        self.assertEqual(
            colormap(dict(values=[0, 1, 2, 3], no_data_value=2)).tolist(),
            [[127, 000, 000, 255],
             INVALID,
             MASKED,
             MASKED],
        )
        # MaskedArray
        self.assertEqual(
            colormap(np.ma.masked_equal([0, 1, 2, 3], 2)).tolist(),
            [[127, 000, 000, 255],
             INVALID,
             MASKED,
             MASKED],
        )
        # floats should not cause errors, not even nan
        floats = -float('inf'), 0.5, 1.5, float('inf'), float('nan')
        result = tuple(colormap(floats).tolist())
        expected = MASKED, [127, 0, 0, 255], INVALID, MASKED, MASKED
        assert expected == result

    def test_data_repr(self):
        data = core.Data(data=(2, 3), limits=(1, 4))
        self.assertEqual(repr(data), '<Data: data 2,3; limits 1,4>')

    def test_register(self):
        name = 'test'
        colormap = discrete()
        colormap.register(name)
        self.assertEqual(colormap, colormaps.get(name))

    def test_not_registered(self):
        self.assertRaises(NameError, colormaps.get, 'nonsense')


class TestManager(unittest.TestCase):
    def test_manager(self):
        manager = colormaps.Manager()
        # check for duplicate keys
        self.assertEqual(len(manager.keys), len(manager.registered))
        # load from disk
        colormap = manager.get('lc-wss')  # load path
        self.assertIsInstance(colormap, core.DiscreteColormap)
        # return from cache
        colormap = manager.get('lc-wss')  # cache path
        self.assertIsInstance(colormap, core.DiscreteColormap)
        # labeling with existing labels
        self.assertEqual(colormap.label([1]),
                         ['1 - BAG - Overig / Onbekend'])
        # not existing colormap
        self.assertRaises(NameError, manager.get, 'blabla')
        # not existing colormap, not existing labels
        colormap = manager.get('jet')
        self.assertEqual(colormap.label([5, 6]), [5, 6])


class TestUtils(unittest.TestCase):
    def test_cdict2config(self):
        n08, n18, n20 = 8, 18, 20
        config = {
            'type': u'GradientColormap',
            'data': [(0.000, [000, 000, 127, 255]),
                     (0.110, [000, 000, 255, 255]),
                     (0.125, [000, 000, 255, 255]),
                     (0.340, [000, 219, 255, 255]),
                     (0.350, [000, 229, 246, 255]),
                     (0.375, [n20, 255, 226, 255]),
                     (0.640, [238, 255, n08, 255]),
                     (0.650, [246, 245, 000, 255]),
                     (0.660, [255, 236, 000, 255]),
                     (0.890, [255, n18, 000, 255]),
                     (0.910, [231, 000, 000, 255]),
                     (1.000, [127, 000, 000, 255])],
        }
        self.assertEqual(utils.cdict2config(cdict()), config)

    def test_save(self):
        name = 'test'
        path = tempfile.mkdtemp()
        here = os.getcwd()
        os.chdir(path)
        utils.save(cdict=cdict(), name=name)
        os.chdir(here)
        self.assertTrue(
            os.path.exists(os.path.join(path, '{}.json'.format(name))),
        )
        shutil.rmtree(path)


def test_create_not_a_dict():
    with raises(TypeError, match='must be a dict'):
        colormaps.create(None)


def test_create_no_type():
    with raises(KeyError, match="must contain 'type'"):
        colormaps.create({})


def test_create_wrong_type():
    with raises(ValueError, match="must be one of"):
        colormaps.create({"type": "UnknownColormap"})


class TestValidation(unittest.TestCase):
    def setUp(self):
        self.kwargs = {
            "data": [[0, BLACK], [1, BLACK]],
            "validate": True,
        }

    def test_discrete_wrong_color(self):
        self.kwargs["data"][0][1] = [0, 0, 0, 0.0]  # contains a float
        with raises(ValueError, match="Invalid color"):
            DiscreteColormap(**self.kwargs)

    def test_discrete_not_unique(self):
        self.kwargs["data"] = [[0, BLACK], [0, BLACK]]  # not unique
        with raises(ValueError, match="Duplicate values in data"):
            DiscreteColormap(**self.kwargs)

    def test_discrete_wrong_masked(self):
        self.kwargs["masked"] = [0, 0, 0, 0.0]  # contains a float
        with raises(ValueError, match="Invalid color"):
            DiscreteColormap(**self.kwargs)

    def test_discrete_wrong_invalid(self):
        self.kwargs["invalid"] = [0, 0, 0, 0.0]  # contains a float
        with raises(ValueError, match="Invalid color"):
            DiscreteColormap(**self.kwargs)

    def test_discrete_negative(self):
        self.kwargs["data"] = [[-1, BLACK]]  # negative
        with raises(ValueError, match="Data values cannot be negative"):
            DiscreteColormap(**self.kwargs)

    def test_gradient_wrong_color(self):
        self.kwargs["data"][0][1] = [0, 0, 0, 0.0]  # contains a float
        with raises(ValueError, match="Invalid color"):
            GradientColormap(**self.kwargs)

    def test_gradient_not_sorted(self):
        self.kwargs["data"] = [[1, BLACK], [0, BLACK]]  # not sorted
        with raises(ValueError, match="'data' must be sorted"):
            GradientColormap(**self.kwargs)

    def test_gradient_log_no_bool(self):
        self.kwargs["log"] = None  # not a bool
        with raises(AssertionError, match="'log' must be a bool"):
            GradientColormap(**self.kwargs)

    def test_gradient_free_no_bool(self):
        self.kwargs["free"] = None  # not a bool
        with raises(AssertionError, match="'free' must be a bool"):
            GradientColormap(**self.kwargs)

    def test_gradient_data_not_sorted(self):
        self.kwargs["data"] = [[1, BLACK], [0, BLACK]]  # not sorted
        with raises(ValueError, match="'data' must be sorted"):
            GradientColormap(**self.kwargs)

    def test_gradient_interp_x_not_sorted(self):
        self.kwargs["interp"] = [[1, 0], [0, 1]]  # x not sorted
        with raises(ValueError, match="x-values in interp must be sorted."):
            GradientColormap(**self.kwargs)

    def test_gradient_interp_y_not_sorted(self):
        self.kwargs["interp"] = [[0, 1], [1, 0]]  # y not sorted
        with raises(ValueError, match="y-values in interp must be sorted."):
            GradientColormap(**self.kwargs)

    def test_gradient_wrong_masked(self):
        self.kwargs["masked"] = [0, 0, 0, 0.0]  # contains a float
        with raises(ValueError, match="Invalid color"):
            GradientColormap(**self.kwargs)


colormaps/core.py
# -*- coding: utf-8 -*-
# (c) Nelen & Schuurmans, see LICENSE.rst.

import numpy as np

MASKED = 0, 0, 0, 0
INVALID = 0, 0, 0, 0

registered = {}


def validate_color(color):
    if len(color) != 4 or not all(
        0 <= e < 256 and isinstance(e, int) for e in color
    ):
        raise ValueError(f"Invalid color: {color}")


class Data(object):
    """
    Convenience wrapper for data.

    scale() will typically be called with the colormaps' limits as the
    clip limits.
    """
    def __repr__(self):
        template = '<{name}: data {dmin},{dmax}; limits {lmin},{lmax}>'
        return template.format(lmin=self.limits[0],
                               lmax=self.limits[1],
                               dmin=self.data.min(),
                               dmax=self.data.max(),
                               name=self.__class__.__name__)

    def __init__(self, data, limits=None):
        """
        If limits, clip data right from start.

        :param data: a numpy data
        :param limits: min, max tuple
        """
        if limits is None:
            self.data = np.asarray(data)  # does not copy if data is array
            self.limits = np.array([self.data.min(), self.data.max()])
        else:
            if limits[1] < limits[0]:
                limits = limits[::-1]
            self.limits = np.array(limits)
            self.data = np.asarray(data).clip(*self.limits)

    def do(self, function):
        """ Apply func to both array and limits. """
        return self.__class__(data=function(self.data),
                              limits=function(self.limits))

    def log(self):
        """ Transform to log domain. """
        return self.do(lambda x: np.log(x))

    def interp(self, interp):
        """ Interpolate. """
        return self.do(lambda x: np.interp(x, *interp))

    def scale(self):
        """ Scale limits and data to span (0, 1). """
        factor = self.limits[1] - self.limits[0]
        if factor == 0:
            # single value, return 0.5
            return self.do(lambda x: np.full(x.shape, 0.5, dtype='f8'))
        offset = self.limits[0]

        # scale
        return self.do(lambda x: (x - offset) / factor)


class BaseColormap(object):
    """ Basic stuff """
    colormap_type = 'base'

    def register(self, name):
        """ Register a colormap for use with get(). """
        registered[name] = self

    def label(self, data, locale=None):
        """ Return a list of labels. """
        if not self.labels:
            return data
        labels = self.labels.get(locale, next(iter(self.labels.values())))
        return np.vectorize(
            lambda x: labels.get(x, x), otypes=[object],
        )(data).tolist()

    def __len__(self):
        """
        Return the official length of the rgba array.

        The actual rgba is one element bigger, for different reasons. The
        DiscreteColormap uses it for storing the 'masked' color. The
        GradientColormap uses it so that the highest value, that is 1.0
        after scaling, does not fall out of the rgba array when looking
        up by index.
        """
        return len(self.rgba) - 1

    def __call__(self, data, limits=None):
        """
        Return rgba array, handle masked values.

        :param data: dict('values': np.array, 'no_data_value': number)
        :param limits: m, n tuple having m <= n.

        The effect of the limits parameter dependes on the colormap type
        """
        if isinstance(data, dict):
            values = np.asarray(data['values'])
            mask = np.equal(values, data['no_data_value'])
        elif isinstance(data, np.ma.MaskedArray):
            values = data.data
            mask = data.mask
        else:
            values = np.asarray(data)
            mask = None

        # also mask non-finite values (-inf, inf, nan)
        nan_mask = ~np.isfinite(values)
        mask = nan_mask if mask is None else mask | nan_mask

        # also mask values outside log's domain
        if getattr(self, "log", False):
            log_mask = values <= 0
            mask = mask | log_mask

        if not np.any(mask):  # np.any(None) returns False
            return self.convert(values, limits)
        rgba = self.masked * np.ones(values.shape + (4,), 'u1')
        if not mask.all():
            rgba[~mask] = self.convert(values[~mask], limits)
        return rgba


class DiscreteColormap(BaseColormap):
    """ Colormap for classified data. """
    colormap_type = 'discrete'

    def __repr__(self):
        template = '<{name}: size {size}, limits {lower}-{upper}>'
        return template.format(size=len(self),
                               lower=self.limits[0],
                               upper=self.limits[1],
                               name=self.__class__.__name__)

    def __init__(self, data,
                 masked=MASKED, invalid=INVALID, labels={}, validate=False):
        """Return colormap instance.

        :param data: list of (value, color)-tuples
        :param masked: color to use for 'no data'
        :param invalid: color to use for values not in data
        :param labels: dict defining locale-specific descriptions

        The values in data must be unique, nonnegative ints. The colors are
        specified as (r, g, b, a) tuples, with each component having an integer
        value ranging from 0 to 255.
        """
        values, colors = zip(*data)
        self.limits = min(values), max(values)

        if validate:
            for color in colors:
                validate_color(color)
            if len(values) != len(set(values)):
                raise ValueError("Duplicate values in data")
            if self.limits[0] < 0:
                raise ValueError("Data values cannot be negative")
            validate_color(masked)
            validate_color(invalid)

        self.masked = np.array(masked, dtype='u1')
        self.invalid = np.array(invalid, dtype='u1')
        self.labels = {k: dict(v) for k, v in labels.items()}
        self.rgba = self.invalid * np.ones((self.limits[1] + 2, 4), dtype='u1')
        self.rgba[-1] = self.masked
        self.rgba[np.array(values)] = colors

    def convert(self, data, limits):
        """"
        Return rgba.

        :param data: A numpy array, not masked.
        :param limits: m, n tuple having m <= n.

        Values outside the limits will be colored with the masked color,
        even if a value was given during the colormap initialization.
        """
        # tighten limits to colormap
        if limits is None:
            limits = self.limits
        else:
            limits = np.array(limits).clip(*self.limits)

        # mask data outside limits
        index = data.astype(np.int64)
        out_of_bounds = (index < limits[0]) | (index > limits[1])
        if index.ndim == 0:  # these cannot be indexed
            if out_of_bounds:
                index = self.limits[1] + 1
        else:
            index[out_of_bounds] = self.limits[1] + 1
        return self.rgba[index]

    def get_legend_data(self, limits, steps):
        """For a discrete map, we ignore 'steps'."""
        if limits is None:
            limits = self.limits
        else:
            # tighten custom limits to colormap
            limits = np.array(limits).clip(*self.limits)

        return np.in1d(
            self.rgba.view('u4')[int(limits[0]):int(limits[1]) + 1],
            self.invalid.view('u4'), invert=True,
        ).nonzero()[0] + limits[0]


class GradientColormap(BaseColormap):
    colormap_type = 'gradient'

    def __repr__(self):
        template = ('<{name}: size {size}, '
                    'limits {lower} - {upper}, log {log}, interp {interp}>')
        return template.format(log=self.log,
                               size=len(self),
                               lower=self.limits[0],
                               upper=self.limits[1],
                               interp=self.interp is not None,
                               name=self.__class__.__name__)

    def process(self, data, limits=None):
        """
        Return processed data.

        If limits are not given, data is scaled according to the colormaps
        interpolation, if any, or else to the input limits.

        Returned data is clipped between 0 and 1.
        """
        data = Data(data=data, limits=limits)
        if self.interp:
            data = data.interp(self.interp)
        if self.log:
            data = data.log()
        data = data.scale()

        # return the clipped array
        return data.data.clip(0, 1)

    def __init__(self, data,
                 size=256, log=False, free=True,
                 interp=None, masked=MASKED, labels={}, validate=False):
        """Return colormap instance.

        :param data: list of (value, color)-tuples that defines the stops
        :param size: size of the generated look-up table
        :param free: use data limits instead of colormap limits
        :param log: use a log scale whenever appropriate
        :param interp: [(x1, y1), (x2, y2), ...]
        :param masked: color


        The values in data must be sorted. The colors are specified as (r, g,
        b, a) tuples, with each component having an integer value ranging from
        0 to 255.

        The colormap constructs a look-up table that distributes its entries
        evenly over the interval spanned by the values in the data parameter.

        In some cases it is more desirable to have finer control over how the
        look-up table entries are distributed. Consider for example the
        elevation of The Netherlands, where it is desirable to have at the same
        time more detail in the densely populated areas near the coast, where
        the elevation is within a few meters from sea level, and also in the
        higher grounds where the elevation is more on the order of 100 meter.

        To achieve this, a transformation can be defined using the `interp`
        parameter, that is applied to both to the values in the `data`
        parameter and to the input data to be colored.

        In a similar fashion, `log=True` can be supplied to apply a log
        function to the values and the input data to achieve a better
        distribution in inputs that have an exponential nature.
        """
        values, colors = zip(*data)

        if validate:
            for color in colors:
                validate_color(color)
            if values != tuple(sorted(values)):
                raise ValueError("'data' must be sorted.")
            assert isinstance(free, bool), "'free' must be a bool."
            assert isinstance(log, bool), "'log' must be a bool."
            if interp is not None:
                interp_x, interp_y = zip(*interp)
                if interp_x != tuple(sorted(interp_x)):
                    raise ValueError("x-values in interp must be sorted.")
                if interp_y != tuple(sorted(interp_y)):
                    raise ValueError("y-values in interp must be sorted.")
            validate_color(masked)

        # options
        self.log = log
        self.free = free
        self.masked = np.array(masked, 'u1')
        self.labels = {k: dict(v) for k, v in labels.items()}
        self.limits = min(values), max(values)

        # store interpolation inputs
        if interp is not None:
            xp, yp = list(np.array(interp, dtype='f8').transpose())
            # np.interp assumes sorted x values
            sort_inds = np.argsort(xp)
            self.interp = xp[sort_inds], yp[sort_inds]
        else:
            self.interp = None

        # now the color entries
        stops = self.process(data=values)
        values = np.arange(size) / (size - 1)

        # build the color table
        self.rgba = np.empty((size + 1, 4), 'u1')
        for i, c in enumerate(zip(*colors)):
            self.rgba[: -1, i] = np.interp(values, stops, c)
        self.rgba[-1, :] = self.rgba[-2, :]

    def convert(self, data, limits):
        """
        Return rgba.

        :param data: A numpy array, not masked.
        :param limits: m, n tuple having m <= n.

        If limits are given, input is scaled such that values between
        limits span the colormap. Values outside are colored according
        to the first and last entries in the colormap.

        If limits are not given, limits are taken from input if the
        colormap is 'free' or else from the colormap initialization
        values.
        """
        if limits is None and not self.free:
            limits = self.limits

        data = self.process(data=data, limits=limits)

        # data is between 0 and 1, and self.rgba deliberately has a repeated
        # last element to handle data elements that have the value 1
        return self.rgba[np.int64(len(self) * data)]

    def get_legend_data(self, limits, steps):
        """We need to interpolate the range, then take a linear range between
        those interpolated values, then extrapolate the range back.

        If this is a log colormap, first log the range, then exp the results
        back."""
        if limits is None:
            limits = self.limits

        if self.log:
            limits = np.log(limits)

        if self.interp:
            interpolated = np.interp(limits, self.interp[0], self.interp[1])
            linear = np.linspace(interpolated[0], interpolated[1], steps)
            data = np.interp(linear, self.interp[1], self.interp[0])
        else:
            data = np.linspace(limits[0], limits[1], steps)

        if self.log:
            data = np.exp(data)

        return data


def get(name):
    try:
        return registered[name]
    except KeyError:
        raise NameError("'{}' is not in registered colormaps".format(name))


def create(colormap, validate=False):
    """ Create a colormap from a dictionary. """
    if not isinstance(colormap, dict):
        raise TypeError("A colormap definition must be a dict")

    try:
        name = colormap["type"]
    except KeyError:
        raise KeyError("A colormap definition must contain 'type'")

    try:
        cls = COLORMAP_TYPES[name]
    except KeyError:
        error = f"'type' must be one of {COLORMAP_TYPES}, not {name}"
        raise ValueError(error)

    kwargs = {"validate": validate, **colormap}
    del kwargs["type"]
    return cls(**kwargs)


COLORMAP_TYPES = {
    DiscreteColormap.__name__: DiscreteColormap,
    GradientColormap.__name__: GradientColormap,
}
